### PR TITLE
SIGSEGV support. Includes:

### PR DESCRIPTION
**Pull request by briankoco**
- raise SIGSEGV from mmap fault handler
 - add ucontext save/restore to signal delivery/return code
 - update signal runtime test with SIGSEGV test

### CODE REVIEW

**Comment by briankoco** [comment]
This gets java hello world to run on JDK 9, which relies on sigsegv to boot the JVM.
It also gets us part of the way to spring boot (#479) but there are other issues still

File: src/unix/system_structs.h [hunk 1]
```diff
@@ -276,6 +276,24 @@ struct rlimit {
...
 #define SIGINFO_UNION_ALIGN 16
 
+/*
+ * SIGSEGV si_codes
+ */
+#define SEGV_MAPERR 1   /* address not mapped to object */
+#define SEGV_ACCERR 2   /* invalid permissions for mapped object */
+//#ifdef __bfin__
+//# define SEGV_STACKFLOW 3   /* stack overflow */
+//#else
```
Comment (wjhun) [inline src/unix/system_structs.h]: Cruft?

File: src/unix/system_structs.h [hunk 2]
```diff
@@ -448,6 +468,113 @@ typedef struct {
...
+    u64 rbp;
+    u64 rbx;
+    u64 rdx;
+    u64 rax;
+    u64 rcx;
+    u64 rsp;
+    u64 rip;
+    u64 eflags; /* RFLAGS */
+    u16 cs;
+    u16 fs;
```
Comment (wjhun) [inline src/unix/system_structs.h]: Not sure what the correct source of this struct is, and we're just filling them with zero anyway, but the order of fs and gs is reversed here vs. linux arch/x86/include/uapi/asm/sigcontext.h.

The comments in that header are somewhat confusing - does Linux actually use these fields for anything meaningful?

File: src/unix/signal.c [hunk 3]
```diff
@@ -298,18 +298,96 @@ static inline void reset_sigmask(thread t)
...
+    mcontext->eflags = f[FRAME_FLAGS];
+    mcontext->cs = f[FRAME_CS];
+    mcontext->fs = 0;
+    mcontext->gs = 0;
+    mcontext->ss = 0; /* FRAME[SS] if UC_SIGCONTEXT SS */
+    mcontext->err = f[FRAME_ERROR_CODE];
+    mcontext->trapno = f[FRAME_VECTOR];
+    mcontext->oldmask = sa->sa_mask.sig[0];
+
+    /* XXX: this should probably come from somewhere in the
```
Comment (wjhun) [inline src/unix/signal.c]: Yes, or feel free to add an entry in frame.h if it makes sense to do so.

File: src/unix/mmap.c [hunk 4]
```diff
@@ -48,34 +65,39 @@ boolean unix_fault_page(u64 vaddr, context frame)
         return true;
     } else {
         /* page protection violation */
+        vmap vm = (vmap)rangemap_lookup(p->vmaps, vaddr);
```
[the remark this answers is not in the document]
Comment (briankoco) [reply src/unix/mmap.c]: makes sense, updated

File: src/unix/system_structs.h (hunk 2 above)
Comment (briankoco) [reply src/unix/system_structs.h]: Nice catch --- don't know how that happened ...

**Pull request merged**